Trigger db commit event after event commit

GrzegorzDrozd opened the pull request:
Hello!

*  Type:new feature

**In raising this pull request, I confirm the following:**

-  [ ] I have read and understood the [Contributing Guidelines](https://github.com/phalcon/phalcon/blob/master/CONTRIBUTING.md)
-  [ ] I have checked that another pull request for this purpose does not exist
-  [ ] I wrote some tests for this PR
-  [ ] I have updated the relevant CHANGELOG
-  [ ] I have created a PR for the [documentation](https://github.com/phalcon/docs) about this change

Small description of change:

Notify the events manager when a transaction is committed or a savepoint is released. This enhances observability and allows external systems to react to these database operations.

Thanks

GrzegorzDrozd committed:
Add event notifications for transaction commits and savepoints

Notify the events manager when a transaction is committed or a savepoint is released. This enhances observability and allows external systems to react to these database operations.

GrzegorzDrozd committed:
Handle different error modes

noone-silent commented on src/Db/Adapter/Pdo/AbstractPdo.php:
```diff
@@ -198,7 +198,13 @@ public function commit(bool $nesting = true): bool
+            /**
+             * Notify the events manager about the committed transaction
+             */
+            $this->fireManagerEvent('db:transactionCommited');
```
Should be `"db:transactionCommitted"`

noone-silent commented on src/Db/Adapter/Pdo/AbstractPdo.php:
```diff
@@ -229,8 +235,14 @@ public function commit(bool $nesting = true): bool
+        $result = $this->releaseSavepoint($savepointName);
+
+        /**
+         * Notify the events manager about the committed savepoint
```
`Notify the events manager about the released savepoint`

GrzegorzDrozd committed:
Handle different error modes

GrzegorzDrozd committed:
Handle different error modes

Jeckerson commented:
Nice, would be great to simulate with tests.

GrzegorzDrozd committed:
Add one more event when a transaction is rolled back
Add unit tests

GrzegorzDrozd commented:
@niden Hi, I see your https://github.com/phalcon/phalcon/pull/628 but I also made one with some additional changes. Can you take a look ?

niden commented:
@GrzegorzDrozd 

This looks good and complements #628 which I will close.

- Please rebase with the current v6.0.0 because the actions do not run without that change.
- You have removed the `commitTransaction` event and replaced it with `committedTransaction`. This will most likely break some apps that utilize the former. Please add the `commitTransaction` as it were and we can have the `committedTransaction` as the final result. Doing so will increase the events triggered but it will allow for an easier transition to v6. Later on we can remove the `commitTransaction`. The same of course for the `rollback`.
- Finally, you mention about a "silent mode" during commit in your comment. How is that detected? By silent mode I assume you mean the setting of `PDO::ATTR_ERRMODE`. If my assumption is correct then you are not checking it there.

niden commented on src/Db/Adapter/Pdo/AbstractPdo.php:
```diff
@@ -198,7 +198,22 @@ public function commit(bool $nesting = true): bool
+            $result = $this->pdo->commit();
+
+            /**
+             * When error mode is set to silent or warning, we need to check result and trigger event only when
```
This is the "silent" mode I refer in my comment

GrzegorzDrozd replied:
Yes, it is about `PDO::ATTR_ERRMODE` value that can be one of those three:

1. PDO::ERRMODE_EXCEPTION - we don't need to check anything, execution will stop
2. PDO::ERRMODE_WARNING - warning will be generated and called pdo method will return false
3. PDO::ERRMODE_SILENT - there will be no warning but called pdo method will return false.

So in cases 2 and 3 we just need to check return from the method. We don't need to know which mode it is.
In case error mode is in exception then event will not be triggered anyway.

GrzegorzDrozd commented:
Hi @niden ,

> Please rebase with the current v6.0.0 because the actions do not run without that change.

OK, will do asap.

> You have removed the `commitTransaction` event and replaced it with `committedTransaction`. This will most likely break some apps that utilize the former. Please add the `commitTransaction` as it were and we can have the `committedTransaction` as the final result. Doing so will increase the events triggered but it will allow for an easier transition to v6. Later on we can remove the `commitTransaction`. The same of course for the `rollback`.

commitTransaction is still there in line 194. I have not removed it. I also did not removed `releaseSavepoint` from line 241 and `rollbackTransaction` from 820. I added new events for that reason - to not break compatibility. I also think that those should be eventually deprecated.

> Finally, you mention about a "silent mode" during commit in your comment. How is that detected? By silent mode I assume you mean the setting of `PDO::ATTR_ERRMODE`. If my assumption is correct then you are not checking it there.

Explained in my other comment. There is no need to read error mode, we just need to check response.

niden commented:
@GrzegorzDrozd Thanks for the explanations. 

This looks good once the tests pass. You will need to run `composer cs-fix` to fix the issue with the test (see run) or just fix it manually.

GrzegorzDrozd committed:
Cs fixes

GrzegorzDrozd commented:
@niden I am not sure what happened but I synced my fork (using github) then:
```shell
> git fetch origin
> git rebase refs/heads/v6.0.x
Rebasing (1/5)
Rebasing (2/5)
Rebasing (3/5)
Rebasing (4/5)
Rebasing (5/5)
Successfully rebased and updated refs/heads/trigger-db-commit-event-after-event-commit.
> git fetch origin
> git merge origin/trigger-db-commit-event-after-event-commit
> git add tests/database/Db/Adapter/Pdo/EventsTest.php
> git commit 
```
I am not that fluent with rebasing :/ Not sure why it shows all those files are changed...
(i used phpstorm for those commands).

niden commented:
Solid work. Thank you @GrzegorzDrozd 

I rebased your branch and force pushed upstream and that solved the issue you were looking at with many files vs. 2.
